File: identity/node/src/webhooks.ts
import { BaseClient } from './client.js'
import type {
  CreateWebhookRequest,
  WebhookCreatedResponse,
  Webhook,
  WebhookListResponse,
  UpdateWebhookRequest,
  RotateSecretResponse,
  WebhookDeliveryListResponse,
} from './types.js'

export class WebhooksResource extends BaseClient {

  /**
   * Register a webhook endpoint to receive signed event callbacks.
   * The `secret` in the response is shown **only once** — store it securely.
   *
   * @example
   * const wh = await kernaq.webhooks.create({
   *   url:    'https://your-server.com/webhooks/kernaq',
   *   events: ['verification.completed', 'verification.failed'],
   * })
   * console.log(wh.secret) // store this — not shown again
   */
  create(req: CreateWebhookRequest): Promise<WebhookCreatedResponse> {
    return this.request<WebhookCreatedResponse>('POST', '/webhooks', req)
  }

  /** List all registered webhooks. Secrets are never returned. */
  list(): Promise<WebhookListResponse> {
    return this.request<WebhookListResponse>('GET', '/webhooks')
  }

  /** Get a single webhook by ID. */
  get(id: string): Promise<Webhook> {
    return this.request<Webhook>('GET', `/webhooks/${id}`)
  }

  /** Update a webhook's URL, events, or active status. All fields optional. */
  update(id: string, req: UpdateWebhookRequest): Promise<Webhook> {
    return this.request<Webhook>('PATCH', `/webhooks/${id}`, req)
  }

  /** Delete a webhook. */
  delete(id: string): Promise<{ message: string }> {
    return this.request<{ message: string }>('DELETE', `/webhooks/${id}`)
  }

  /**
   * Rotate the signing secret for a webhook.
   * The new secret is active immediately. Store it before this call returns.
   */
  rotateSecret(id: string): Promise<RotateSecretResponse> {
    return this.request<RotateSecretResponse>('POST', `/webhooks/${id}/rotate-secret`)
  }

  /** List the last 50 delivery attempts for a webhook — useful for debugging. */
  listDeliveries(id: string): Promise<WebhookDeliveryListResponse> {
    return this.request<WebhookDeliveryListResponse>('GET', `/webhooks/${id}/deliveries`)
  }
}

File: identity/go/kernaq/webhooks.go
package kernaq

import "context"

// WebhooksResource handles /v1/webhooks.
type WebhooksResource struct {
	c *client
}

// Create registers a webhook endpoint to receive signed event callbacks.
// The Secret in the response is shown only once — store it securely.
func (r *WebhooksResource) Create(ctx context.Context, req CreateWebhookRequest) (*WebhookCreatedResponse, error) {
	var out WebhookCreatedResponse
	if err := r.c.postJSON(ctx, "/webhooks", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// List returns all registered webhooks for the project. Secrets are never returned.
func (r *WebhooksResource) List(ctx context.Context) (*WebhookListResponse, error) {
	var out WebhookListResponse
	if err := r.c.get(ctx, "/webhooks", &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get returns a single webhook by ID.
func (r *WebhooksResource) Get(ctx context.Context, id string) (*Webhook, error) {
	var out Webhook
	if err := r.c.get(ctx, "/webhooks/"+id, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Update updates a webhook's URL, events, or active status.
func (r *WebhooksResource) Update(ctx context.Context, id string, req UpdateWebhookRequest) (*Webhook, error) {
	var out Webhook
	if err := r.c.patchJSON(ctx, "/webhooks/"+id, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Delete removes a webhook.
func (r *WebhooksResource) Delete(ctx context.Context, id string) error {
	return r.c.do(ctx, "DELETE", "/webhooks/"+id, nil, "", nil, nil)
}

// RotateSecret generates a new signing secret for a webhook.
// The new secret is returned once and active immediately.
func (r *WebhooksResource) RotateSecret(ctx context.Context, id string) (*RotateSecretResponse, error) {
	var out RotateSecretResponse
	if err := r.c.do(ctx, "POST", "/webhooks/"+id+"/rotate-secret", nil, "", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ListDeliveries returns the last 50 delivery attempts for a webhook.
func (r *WebhooksResource) ListDeliveries(ctx context.Context, id string) (*WebhookDeliveryListResponse, error) {
	var out WebhookDeliveryListResponse
	if err := r.c.get(ctx, "/webhooks/"+id+"/deliveries", &out); err != nil {
		return nil, err
	}
	return &out, nil
}
